enhancements/security/manual-cloud-credentials.md

```diff
@@ -0,0 +1,170 @@
+
+However if CCO is disabled, it will compare the ClusterVersion.Spec.AvailableUpdates to this list of ReconciledAvailableUpdates. If *any* available update is not present in the reconciled list, the CCO will report Upgradable=False.
+
+### Initial Installation
+
+```bash
+$ oc adm credentials reconcile --manifests=/path/to/installer/manifests
+```
+
+This command would parse the CredentialsRequests in the manifests directory, mint credentials in the cloud using the local users environment variables or credentials file, store the status on the CredentialsRequsts (which are yet to be created), write Secret manifests containing the cloud credentials, and disable the CCO.
+
+### Upgrade Reconciliation
+
+```bash
+$ oc adm credentials reconcile --available-updates
+```
+
+This command will iterate all currently available updates in ClusterVersion.AvailableUpdates. For each given release image we will extract each CredentialsRequest and use the local users environment variables/credentials file to compare and adjust permissions as necessary.
+
+This command reconciles permissions *in an additive fashion*. We do not want to reconcile any reduction in permissions as that would affect the running components in the cluster right now. The command would print a warning if this situation is encountered, and it can be resolved after upgrade with the next command.
+
+For net new CredentialsRequests in the update, we have a problem. We cannot assume to store a secret in the target namespace as it may be for a component that does not exist in the currently running OpenShift release. To work around this problem we propose copying net new CredentialsRequests into the cluster (all CredentialsRequests live in the openshift-cloud-credential-operator namespace today), and introducing the concept of temporary secret storage. This would allow us to mint the credentials when we have the root creds available, grab the secrets which we typically only get at creation time and cannot retrieve later, and transfer them to the correct namespace when it is available.
+
+Each available update we process successfully is added to the reconciledAvailableUpdates list.
+
+### On-Going Reconciliation
+
+```bash
+$ oc adm credentials reconcile
+```
+
+This command would re-reconcile against the CredentialsRequests currently in the cluster. This is more of an on-going configuration management to ensure something doesn't change unexpectedly on us due to external factors. It would however support reducing permissions as a result of an additive only reconcile pre-upgrade.
+
+It should be noted that running this command after reconciling --available-updates, but before updating, would undo the pre-upgrade changes. As such this command should clear the reconciledAvailableUpdates list.
+
+
+## Open Questions [optional]
+
+ > 1. How can we handle net new CredentialsRequests in incoming release images?
```

Comment:
what happens in an rhacm context?  is there anything we can have hive build on top of this to make supporting this use case at scale better?  if i manage 100+ clusters and dont want to give you power to mint creds, what would you recommend?

Reply:
It's been ages since we talked about this, way back to when first designing CCO, but we always hoped it could be leveraged in a centralized Hive cluster to keep root creds out of the end cluster. My first instinct would be that this is a separate feature and enhancement but I will take a pass through this and see if there's anything here that we could reuse in such a scenario so we're better prepared for it.